File: day01/day1.ml
(*
let read_file filename =
let lines = ref [] in
let chan = open_in filename in
try
  while true; do
    lines := input_line chan :: !lines
  done; !lines
with End_of_file ->
  close_in chan;
  List.rev !lines ;;

let part1 =
    let contents = read_file "data.txt" in
    let parsed = List.map (fun x -> List.filter (fun y -> y <> "") (String.split_on_char ' ' x)) contents in
    List.map (fun row -> print_endline (List.fold_left (fun acc x -> x ^ ", " ^ acc) "" row)) parsed;
    let xs = List.map (fun x -> int_of_string (List.hd x)) parsed in
    let ys = List.map (fun y -> int_of_string (List.hd (List.tl y))) parsed in
    let dist = List.map2 (fun x y -> abs (x - y)) xs ys in
    List.fold_right (+) dist 0;;

let () = print_endline part1;;
*)

File: day01/day1.hs
import Data.List (sort, transpose)

part1 :: IO ()
part1 = do
  contents <- readFile "data.txt"
  let as : bs : _ = map (sort . map read) $ transpose $ map words $ lines contents
      dist = sum $ map abs $ zipWith (-) as bs
  print dist

part2 :: IO ()
part2 = do
  contents <- readFile "data.txt"
  let as : bs : _ = map (map read) $ transpose $ map words $ lines contents
      sim = sum $ map (\a -> a * length (filter (== a) bs)) as
  print sim
